### ISSUE TITLE
Windows: System Certificate Store might fail to find certs by their public key fingerprint

### ISSUE THREAD

**Issue by reneme** [randombit/botan#2779]
While writing tests for our application, I noticed that Botan's `System_Certificate_Store` fails to find certain certificates by their public key SHA-1 ([`Certificate_Store_Windows::find_cert_by_pubkey_sha1`](https://github.com/randombit/botan/blob/master/src/lib/x509/certstor_system_windows/certstor_windows.cpp#L231)) on Windows. For instance `D-TRUST Root Class 3 CA 2 2009` (below) is such a certificate. 

The reason: This certificate does not adhere to [RFC 3280's suggestion](https://datatracker.ietf.org/doc/html/rfc3280#section-4.2.1.2) to calculate the "subject key identifier" as a SHA-1 *without* prepending the algorithm identifier in the BER encoding. Hence, the "subject key identifier" differs from the canonical "public key SHA-1" for this particular certificate. In my understanding, that is not the usual case, though.

```
# D-TRUST Root Class 3 CA 2 2009
Subject Key Identifier: fdda14c49f30de21bd1e4239fcab632349e0f184
Public Key SHA-1:       a737b46280e401211faff74eeccd1c05eb8947ce

# Counter Example: DigiCert Global Root G2
Subject Key Identifier: 4e2254201895e6e36ee60ffafab912ed06178f39
Public Key SHA-1:       4e2254201895e6e36ee60ffafab912ed06178f39
```

Unfortunately, [the Windows API](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certfindcertificateinstore) does not allow to search for a SHA1 of a certificate's public key directly (it only supports MD5 -- via `CERT_FIND_PUBKEY_MD5_HASH`). Instead, Botan must rely on the convention described above to implement finding certificates via their SHA1 public key fingerprint and actually filter for the "subject key identifier" (via `CERT_FIND_KEY_IDENTIFIER`). Hence, certificates like `D-TRUST Root Class 3 CA 2 2009` will not be found via `::find_cert_by_pubkey_sha1` on Windows despite the certificate being in the trust store. For macOS and Linux this issue doesn't appear.

I am opening this ticket mainly for documentation purposes, in case somebody else stumbles over this. Nevertheless, any suggestion how to tackle this would be appreaciated. It remains to be tested, how many root CA certificates suffer from this issue on a commodity desktop operating system installation.

```
-----BEGIN CERTIFICATE-----
MIIEMzCCAxugAwIBAgIDCYPzMA0GCSqGSIb3DQEBCwUAME0xCzAJBgNVBAYTAkRF
MRUwEwYDVQQKDAxELVRydXN0IEdtYkgxJzAlBgNVBAMMHkQtVFJVU1QgUm9vdCBD
bGFzcyAzIENBIDIgMjAwOTAeFw0wOTExMDUwODM1NThaFw0yOTExMDUwODM1NTha
ME0xCzAJBgNVBAYTAkRFMRUwEwYDVQQKDAxELVRydXN0IEdtYkgxJzAlBgNVBAMM
HkQtVFJVU1QgUm9vdCBDbGFzcyAzIENBIDIgMjAwOTCCASIwDQYJKoZIhvcNAQEB
BQADggEPADCCAQoCggEBANOySs96R+91myP6Oi/WUEWJNTrGa9v+2wBoqOADER03
UAifTUpolDWzU9GUY6cgVq/eUXjsKj3zSEhQPgrfRlWLJ23DEE0NkVJD2IfgXU42
tSHKXzlABF9bfsyjxiupQB7ZNoTWSPOSHjRGICTBpFGOShrvUD9pXRl/RcPHAY9R
ySPocq60vFYJfxLLHLGvKZAKyVXMD9O0Gu1HNVpK7ZxzBCHQqr0ME7UAyiZsxGsM
lFqVlNpQmvH/pStmMaTJOKDfHR+4CS7zp+hnUquVH+BGPtikw8paxTGA6Eian5Rp
/hnd2HN8gcqW3o7tszIFZYQ05ub9VxC1X3a/L7AQDcUCAwEAAaOCARowggEWMA8G
A1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFP3aFMSfMN4hvR5COfyrYyNJ4PGEMA4G
A1UdDwEB/wQEAwIBBjCB0wYDVR0fBIHLMIHIMIGAoH6gfIZ6bGRhcDovL2RpcmVj
dG9yeS5kLXRydXN0Lm5ldC9DTj1ELVRSVVNUJTIwUm9vdCUyMENsYXNzJTIwMyUy
MENBJTIwMiUyMDIwMDksTz1ELVRydXN0JTIwR21iSCxDPURFP2NlcnRpZmljYXRl
cmV2b2NhdGlvbmxpc3QwQ6BBoD+GPWh0dHA6Ly93d3cuZC10cnVzdC5uZXQvY3Js
L2QtdHJ1c3Rfcm9vdF9jbGFzc18zX2NhXzJfMjAwOS5jcmwwDQYJKoZIhvcNAQEL
BQADggEBAH+X2zDI36ScfSF6gHDOFBJpiBSVYEQBrLLpME+bUMJm2H6NMLVwMeni
acfzcNsgFYbQDfC+rAF1hM5+n02/t2A7nPPKHeJeaNijnZflQGDSNiH+0LS4F9p0
o3/U37CYAqxva2ssJSRyoWXuJVrl5jLn8t+rSfrzkGkj2wTZ51xY/GXUl77M/C4K
zCUqNQT4YJEVdT1B/yMfGchs64JTBKbkTCJNjYy6zltz7GRUUG3RnFX7acM2w4y8
PIWmawomDeCTmGCufsYkl4phX5GOZpIJhzbNi5stPvZR1FDUWSi9g/LMKHtThm3Y
Johw1+qRzT65ysCQblrGXnRl11z+o+I=
-----END CERTIFICATE-----
```

**Comment by randombit** [comment]
Possible solution: the store could (perhaps on construction) perform a linear scan over the entire trust store and compute the SHA-1 pubkey hash of each cert. For any certs where that value doesn't match the SKID, such as this D-TRUST root, it would store a map from SHA-1 to the certificate SKID. Then `find_cert_by_pubkey_sha1` would examine the map first, if the SHA-1 is listed there then it uses the stored SKID instead of the hash.

Parsing each and every certificate might prove expensive, but it may be tolerable if done just once on first use.

**Comment by reneme** [comment]
I opened a PR that implements a fallback which relies on the Windows API outputs of `CertEnumCertificatesInStore`. Coincidentially, this contains the Public Key in the right format to be hashed and compared to the query parameter. Hence, no full certificate parsing via `X509_Certificate` is necessary. The runtime of this fallback seems to be tolerable.

FWIW, in my macOS certificate store, I found 17 certificates that pose the described issue (`keyID != SHA1(pubkey`). Five of which don't have a "Subject Key Identifier" at all.
